### PR TITLE
Use generic lists in Control

### PR DESCRIPTION
**Pull request by gpetrou**
## Proposed changes

- Use generic lists in Control.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3511)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 1]
```diff
@@ -5855,12 +5853,12 @@ private static ArrayList GetChildWindows(IntPtr hWndParent)
...
-        ///  - Returns a TabIndex sorted array of ControlTabOrderHolder objects.
+        ///  - Returns a TabIndex sorted list of ControlTabOrderHolder objects.
         /// </summary>
-        private ArrayList GetChildWindowsTabOrderList()
+        private List<ControlTabOrderHolder> GetChildWindowsTabOrderList()
         {
-            ArrayList holders = new ArrayList();
-            ArrayList windows = GetChildWindows(Handle);
+            List<ControlTabOrderHolder> holders = new List<ControlTabOrderHolder>();
+            List<IntPtr> windows = GetChildWindows(Handle);
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:5861]: [perf] Swap these around, and use the size of `windows` list to initialise the size of `holders` list to mitigate possible resize requests.

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 2]
```diff
@@ -5855,12 +5853,12 @@ private static ArrayList GetChildWindows(IntPtr hWndParent)
         ///  - Controls with the same TabIndex remain in original relative child index order (= z-order).
         ///  - Child windows with no corresponding Control objects (and therefore no discernable TabIndex)
         ///  are sorted to the front of the list (but remain in relative z-order to one another).
-        ///  - Returns a TabIndex sorted array of ControlTabOrderHolder objects.
+        ///  - Returns a TabIndex sorted list of ControlTabOrderHolder objects.
         /// </summary>
-        private ArrayList GetChildWindowsTabOrderList()
+        private List<ControlTabOrderHolder> GetChildWindowsTabOrderList()
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:5858]: Actually this method has a single call site, which is not interested in the resulting list 
https://github.com/dotnet/winforms/blob/1678a756fb6ac4c7befd712cba1dc0f3cbc3b078/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs#L5803-L5815

So here's a good opportunity to simplify the code, remove redundant methods and reduce allocations.

Comment (gpetrou) [reply src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:5858]: But the holders list is used, right?

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:5858]: Essentially something like this:
```cs
        private int[] GetChildWindowsInTabOrder()
        {
            List<IntPtr> windows = new List<IntPtr>();

            for (IntPtr hWndChild = User32.GetWindow(Handle, User32.GW.CHILD);
                 hWndChild != IntPtr.Zero;
                 hWndChild = User32.GetWindow(hWndChild, User32.GW.HWNDNEXT))
            {
                windows.Add(hWndChild);
            }

            List<ControlTabOrderHolder> holders = new List<ControlTabOrderHolder>(windows.Count);

            foreach (IntPtr hWnd in windows)
            {
                Control ctl = FromHandle(hWnd);
                int tabIndex = (ctl == null) ? -1 : ctl.TabIndex;
                holders.Add(new ControlTabOrderHolder(holders.Count, tabIndex, ctl));
            }

            holders.Sort(new ControlTabOrderComparer());

            int[] indexes = new int[holders.Count];
            for (int i = 0; i < holders.Count; i++)
            {
                indexes[i] = holders[i]._oldOrder;
            }

            return indexes;
        }
```

**Pull request merged**